site/source/docs/api_reference/wasm_audio_worklets.rst

```diff
@@ -0,0 +1,162 @@
+the JS code for the ScriptProcessorNode files in the AudioWorkletGlobalScope.
+This is managed automatically by the Emscripten Wasm AudioWorklets runtime.
+
+Instead, application developers will need to implement a small amount of JS <-> Wasm
+(C/C++) interop to interact with the AudioContext and AudioNodes from Wasm.
+
+Audio Worklets operate on a two layer "class type & its instance" design:
+first one defines one or more node types (or classes) called AudioWorkletProcessors,
+and then, these processors are instantiated one or more times in the audio
+processing graph as AudioWorkletNodes.
+
+Once a class type is instantiated on the Web Audio graph and the graph is
+running, a C/C++ function pointer callback will be invoked for each 128
+samples of the processed audio stream that flows through the node.
+
+This callback will be executed on a dedicated separate audio processing
+thread with real-time processing priority. Each Web Audio context will
+utilize only a single audio processing thread. That is, even if there are
+multiple audio node instances (maybe from multiple different audio processors),
+these will all share the same dedicated audio thread on the AudioContext,
+and will not run in a separate thread of their own each.
+
+Programming Example
+===================
+
+To get hands-on experience with programming Wasm Audio Worklets, let's create a
+simple audio node that outputs random noise through its output channels.
+
+1. First, we will create a Web Audio context in C/C++ code. This is achieved
+via the ``emscripten_create_audio_context()`` function. In a larger application
+that integrates existing Web Audio libraries, you may already have an
+``AudioContext`` created via some other library, in which case you would instead
+register that context to be visible to WebAssembly by calling the function
+``emscriptenRegisterAudioObject()``.
+
+Then, we will instruct the Emscripten runtime to initialize a Wasm Audio Worklet
+thread scope on this context. The code to achieve these tasks looks like:
+
+.. code-block:: cpp
+
+  #include <emscripten/webaudio.h>
+
+  uint8_t audioThreadStack[4096];
+
+  int main()
+  {
+    EMSCRIPTEN_WEBAUDIO_T context = emscripten_create_audio_context(0);
+
+    emscripten_start_wasm_audio_worklet_thread_async(context, audioThreadStack, sizeof(audioThreadStack),
+                                                     &AudioThreadInitialized, 0);
+  }
+
+2. When the worklet thread context has been initialized, we are ready to define our
+own noise generator AudioWorkletProcessor node type:
+
+.. code-block:: cpp
+
+  void AudioThreadInitialized(EMSCRIPTEN_WEBAUDIO_T audioContext, EM_BOOL success, void *userData)
+  {
+    if (!success) return; // Check browser console in a debug build for detailed errors
+    WebAudioWorkletProcessorCreateOptions opts = {
+      .name = "noise-generator",
+    };
+    emscripten_create_wasm_audio_worklet_processor_async(audioContext, &opts, &AudioWorkletProcessorCreated, 0);
```

Comment:
I don't know much about Audio Worklets so I'm trying to read this starting from the docs. It looks like this refers to `AudioWorkletProcessorCreated` before it is defined (that happens on line 103 below). Is that a typo, or intentional? (The same happens for `AudioThreadInitialized` it seems)

And how should I be reading this - does `AudioThreadInitialized` initialize an audio thread, or is it a callback when one is created? How does that relate to `AudioWorkletProcessorCreated` - is there one processor per thread, or one to many, or something else?

Sorry for the basic questions!

Reply:
It is intentional. Of course in a full C++ program a declaration must precede the location of use, but for prose text in steps 1.2.3. order it flows better to first describe a step "2. this is how we create an audio worklet processor", followed by "3. this is how we continue to do when the processor has been created".

The code example here does not intend to be a linearly copy-pasteable sequence of code to one's own program, but the developer will need to organize the code in the right C++ program order.

> does AudioThreadInitialized initialize an audio thread, or is it a callback when one is created?

`AudioThreadInitialized` is the callback that gets called after the audio thread has already been initialized. That thread context is initialized by the function call `emscripten_start_wasm_audio_worklet_thread_async`, so `AudioThreadInitialized` is its async completion callback.

> How does that relate to AudioWorkletProcessorCreated

`AudioWorkletProcessorCreated` callback is fired when a worklet processor is created. Defining worklet processors is unfortunately also an asynchronous operation in the Audio Worklet API.

> is there one processor per thread, or one to many, or something else?

This is referred to on doc lines 46-51: "Each Web Audio context will
 utilize only a single audio processing thread. That is, even if there are
 multiple audio node instances (maybe from multiple different audio processors),  these will all share the same dedicated audio thread on the AudioContext, and will not run in a separate thread of their own each."

i.e. all audio node instances (whether they are instantiated off of the same processor type or each are instances of separate processor types) run in that same singleton audio thread that each Web Audio context has.

If more than one audio thread was desired, one would likely have to create multipled Web Audio contexts, though I think that might go against how the Audio Worklet API would like to be used. I posted an issue at https://github.com/WebAudio/web-audio-api/issues/2500 about this, as someone else asked this before as well.

Reply:
Added a small doc update commit to the end to highlight that AudioWorklets won't allow developers to create synchronously blocking `for(;;)` loops, i.e. it is not a general real-time audio pthread solution.

Reply:
I see, thanks!